dask_expr/_expr.py

```diff
@@ -2923,6 +2838,62 @@ def _execute_task(graph, name, *deps):
         return dask.core.get(graph, name)
 
 
+def collect_depdendents(expr) -> defaultdict:
+    dependents = defaultdict(list)
+    stack = [expr]
+    seen = set()
+    while stack:
+        node = stack.pop()
+        if node._name in seen:
+            continue
+        seen.add(node._name)
+
+        for dep in node.dependencies():
+            stack.append(dep)
+            dependents[dep._name].append(weakref.ref(node))
```

Comment:
This is a little bit of a shortcut, in case we can prune whole branches so that we don't have to consider them when doing optimisations

You could also interpret this as premature optimization, but I was a bit worried that I would keep things alive that will screw me over if I have hard refs

Reply:
I'm not blocking on this but it feels a little brittle to rely on weakref semantics to get this right. I'd be worried if this logic would not work out without weakrefs

Reply:
No the logic works without, it would just need more simplify steps to get to the final result